File: Excel/900-999/930/930 Challenge.R
library(tidyverse)
library(readxl)

path <- "900-999/930/930 Start Stop RunTime.xlsx"
input <- read_excel(path, range = "A2:C22")
test <- read_excel(path, range = "E2:H13")

result = input %>%
  mutate(rn = row_number(), .by = c(Machine, EventType)) %>%
  pivot_wider(names_from = EventType, values_from = EventTime) %>%
  arrange(Machine, rn) %>%
  replace_na(list(Stop = as.POSIXct('2024-03-01 19:00:00'))) %>%
  mutate(RunMinutes = as.numeric(difftime(Stop, Start, units = "mins"))) %>%
  select(Machine, StartTime = Start, StopTime = Stop, RunMinutes)

all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE

File: Excel/900-999/930/930 Challenge.py
import pandas as pd
from datetime import datetime

path = "900-999/930/930 Start Stop RunTime.xlsx"

input = pd.read_excel(path, skiprows=1, usecols="A:C", nrows=20)
test = pd.read_excel(path, skiprows=1, usecols="E:H", nrows=11).rename(columns=lambda c: c.rstrip(".1"))

result = (
    input
    .assign(rn=lambda df: df.groupby(["Machine", "EventType"]).cumcount())
    .pivot(index=["Machine", "rn"], columns="EventType", values="EventTime")
    .reset_index()
    .sort_values(["Machine", "rn"])
    .assign(
        Stop=lambda df: df["Stop"].fillna(datetime(2024, 3, 1, 18, 0, 0)),
        RunMinutes=lambda df: ((df["Stop"] - df["Start"]).dt.total_seconds() / 60).astype("int64"),
    )
    .rename(columns={"Start": "StartTime", "Stop": "StopTime"})
    [["Machine", "StartTime", "StopTime", "RunMinutes"]]
    .reset_index(drop=True)
)

print(result.equals(test))
# True

